src/QMCDrivers/MCPopulation.cpp

```diff
@@ -64,18 +100,40 @@ void MCPopulation::createWalkers(IndexType num_walkers, const WalkerConfiguratio
 
   outputManager.pause();
 
-  //this part is time consuming, it must be threaded and calls should be thread-safe.
-#pragma omp parallel for
+  // nextWalkerID is not thread safe so we need to get our walker id's here
+  // before we enter a parallel section.
+  std::vector<long> walker_ids(num_walkers_plus_reserve, 0);
+  // notice we only get walker_ids for the number of walkers in the walker_configs
+  for (size_t iw = 0; iw < num_walkers; iw++)
+    walker_ids[iw] = nextWalkerID();
+
+  // this part is time consuming, it must be threaded and calls should be thread-safe.
+  // It would make more sense if it was over crowd threads as the thread locality of the walkers
+  // would at least initially be "optimal" Depending on the number of OMP threads and implementation
+  // this may be equivalent.
+#pragma omp parallel for shared(walker_ids)
   for (size_t iw = 0; iw < num_walkers_plus_reserve; iw++)
   {
-    walkers_[iw]             = std::make_unique<MCPWalker>(elec_particle_set_->getTotalNum());
+    int parent_id = 0;
+    walkers_[iw]             = std::make_unique<MCPWalker>(walker_ids[iw], parent_id, elec_particle_set_->getTotalNum() );
     walkers_[iw]->Properties = elec_particle_set_->Properties;
 
     // initialize coord
     if (const auto num_existing_walkers = walker_configs.getActiveWalkers())
+    {
+      // Save this walkers id.
+      auto this_walkers_id = walkers_[iw]->getWalkerID();
+      // the walker assignment operator is basically a simple copy so walker_id is overwritten with
+      // something meaningful only in another QMC section.
       *walkers_[iw] = *walker_configs[iw % num_existing_walkers];
+      // An outside section context parent ID is multiplied by -1.
+      walkers_[iw]->setParentID(-1 * walker_configs[iw % num_existing_walkers]->getWalkerID());
+      walkers_[iw]->setWalkerID(this_walkers_id);
```

Comment:
really bad code. save, overwritten, overwritten.
Is it possible to move the creation of walkers_[iw] in to the branch and construct using copy constructors?

Reply:
I'll look into it this is indeed ugly.

Reply:
Working this a little I don't see a way around reworking walker constructor and id semantics and I think that needs to be a separate PR. I think it will be easier since this will ripple to tests, comments, and I think into files in the next PR in this series.

Reply:
The current code is correct. No problem updating it later as we clean up Walker class constructors.